File: src/test/java/br/com/jiratorio/base/resolver/SpecificationResolver.java
package br.com.jiratorio.base.resolver;

import br.com.jiratorio.base.specification.NotFound;
import io.restassured.builder.ResponseSpecBuilder;
import io.restassured.specification.ResponseSpecification;
import org.apache.http.HttpStatus;
import org.hamcrest.Matchers;
import org.junit.jupiter.api.extension.ExtensionContext;
import org.junit.jupiter.api.extension.ParameterContext;
import org.junit.jupiter.api.extension.ParameterResolutionException;
import org.junit.jupiter.api.extension.ParameterResolver;

public class SpecificationResolver implements ParameterResolver {

    @Override
    public boolean supportsParameter(final ParameterContext parameterContext,
                                     final ExtensionContext extensionContext) throws ParameterResolutionException {
        return parameterContext.isAnnotated(NotFound.class);
    }

    @Override
    public Object resolveParameter(final ParameterContext parameterContext,
                                   final ExtensionContext extensionContext) throws ParameterResolutionException {
        if (parameterContext.isAnnotated(NotFound.class)) {
            return notFoundSpec();
        }

        return null;
    }

    private ResponseSpecification notFoundSpec() {
        return new ResponseSpecBuilder()
                .expectStatusCode(HttpStatus.SC_NOT_FOUND)
                .expectBody("error", Matchers.is("Not Found"))
                .expectBody("message", Matchers.is("Resource Not Found"))
                .expectBody("status", Matchers.is(404))
                .build();
    }

}

File: src/test/java/br/com/jiratorio/base/resolver/SpecificationResolver.kt
package br.com.jiratorio.base.resolver

import br.com.jiratorio.base.specification.NotFound
import io.restassured.builder.ResponseSpecBuilder
import org.apache.http.HttpStatus
import org.hamcrest.Matchers
import org.junit.jupiter.api.extension.ExtensionContext
import org.junit.jupiter.api.extension.ParameterContext
import org.junit.jupiter.api.extension.ParameterResolver

class SpecificationResolver : ParameterResolver {

    override fun supportsParameter(parameterContext: ParameterContext, extensionContext: ExtensionContext) =
            parameterContext.isAnnotated(NotFound::class.java)

    override fun resolveParameter(parameterContext: ParameterContext, extensionContext: ExtensionContext?): Any? {
        if (parameterContext.isAnnotated(NotFound::class.java)) {
            return notFoundSpec()
        }

        return null
    }

    private fun notFoundSpec() = ResponseSpecBuilder()
            .expectStatusCode(HttpStatus.SC_NOT_FOUND)
            .expectBody("error", Matchers.`is`("Not Found"))
            .expectBody("message", Matchers.`is`("Resource Not Found"))
            .expectBody("status", Matchers.`is`(404))
            .build()

}
